### PR TITLE
Add language support for `hosts` files

### LINKED ISSUE #6389: Add Language: Hosts
## Language name

`Hosts`, alternately `Hosts file`.

## URL of example repository

https://github.com/StevenBlack/hosts

## URL of syntax highlighting grammar

https://marketplace.visualstudio.com/items?itemName=tommasov.hosts

## Most popular extensions

A large supermajority seem to use `.txt` suffixes. Some uncommon cases use no suffixes at all, primarily lists with long history that were meant to overwrite operating systems' `/etc/hosts` files.

## Detected language

Currently none.

## Submitter's notes

In relation to the previously approved and implemented ["Adblock Filter Syntax" language](https://github.com/github/linguist/pull/5968), I talked a bit with that syntax's main owners at https://github.com/AdguardTeam/VscodeAdblockSyntax/issues/70, who from what I understood recommended to me https://marketplace.visualstudio.com/items?itemName=tommasov.hosts as the recommended method to add colors to Hosts lists/entries.

["Hosts" files' history](https://en.wikipedia.org/wiki/Hosts_(file)) stretch back to the late 1980's, and have been used for adblocking since the mid-to-late 1990's (["Peter Lowe's List"](https://pgl.yoyo.org/adservers/), ["Dan Pollock's Hosts"](https://someonewhocares.org/hosts/)). It is occasionally used to this day by various tools (uBlock Origin, AdGuard Home, Pi-hole, various smaller tools), although it has never been fully implemented in AdGuard Browser Extension or Adblock Plus.

As with "Adblock Filter Syntax", implementation in GitHub Linguist would rely largely on repos' `.gitattritubes` specifying filetypes and `linguist-detectable`. I estimate a low-5-digit amount of GitHub repos using Hosts syntax, some with ≥1,000 stars.

### PR DESCRIPTION
**Pull request by Alhadis**
## Description
This PR adds language support for [`hosts`](https://en.wikipedia.org/wiki/Hosts_(file)) files, as detailed by @DandelionSprout in #6389.

### Usage
A filename search for `hosts` yields over [~25,200 results][1]. Most of these are legit `hosts(5)` files, but a non-trivial number of them _aren't_ (such as [this](https://github.com/avpatel/xvisor-next/blob/1329959e5c3ab0f1d6b2956e5bb7450cf677b578/HOSTS)). I contemplating registering this a generic extension, but quickly realised this feature only accommodates file-extensions, *not* unabridged filenames (which are handled by two different classification strategies). I'm wondering if the “generic filetypes” feature added to Linguist in 2020 should accommodate “generic filenames” as well. @lildude, any thoughts?

## Checklist
- [x] **I am adding a new language.**
	-	[x] **The extension of the new language is used in hundreds of repositories.**  
		[~25.2k search results][1] for the filename `hosts`.
	-	[x] **I have included a real-world usage sample.**
		-	[`samples/Hosts File/filenames/hosts`][]
			- **Source:** Original
			- **License:** Public domain[^1]
		-	[`samples/INI/filenames/hosts`][]
			- **Source:** [`PacktPublishing/Ansible-for-Real-life-Automation`](https://github.com/PacktPublishing/Ansible-for-Real-life-Automation)[^2]
			- **License:** [MIT](https://github.com/PacktPublishing/Ansible-for-Real-life-Automation/blob/07d9a59932beadf7b92e8c9983cc7349548d105f/LICENSE)
	-	[x] **I have ~~included~~ _updated_ a syntax highlighting grammar.**  
		`langauge-etc` already includes a [grammar][2] for highlighting [`hosts(5)`][] syntax
	-	[x] **I have added a colour**
		- **Hex value:** `#308888`
		- **Rationale:** Arbitrarily chosen from a cold, clinical colour palette that reflects the “bare metal” nature of Internetworking.
	-	[ ] ~~I have updated the heuristics.~~

Closes #6389

[^1]: File is ineligible for copyright, at least according to my reading of [Wikipedia](https://en.wikipedia.org/wiki/Threshold_of_originality).
[^2]: Concatenated from several configuration examples embedded in markdown files.

[`samples/Hosts File/filenames/hosts`]: https://github.com/github/linguist/blob/5b4bbf0d547393f87d81b1ac60d147a375c42ec7/samples/Hosts%20File/filenames/hosts
[`samples/INI/filenames/hosts`]: https://github.com/github/linguist/blob/5b4bbf0d547393f87d81b1ac60d147a375c42ec7/samples/INI/filenames/hosts
[1]: https://github.com/search?q=path%3A**%2Fhosts&type=code
[2]: https://github.com/Alhadis/language-etc/blob/master/grammars/hosts.cson
[`hosts(5)`]: http://man.openbsd.org/hosts.5

### CODE REVIEW

**Comment by Alhadis** [comment]
@lildude Any idea how we can get our tests to pass when both `hosts` _and_ `HOSTS` are expected to exist simultaneously under `samples/Hosts File/filenames/`…? Assuming a case-insensitive filesystem and everything…

(This would be a non-issue if Linguist matched filenames case-insensitively, the way it does with file extensions).

**Review by Mzariez01 — APPROVED** [review]

**Review by justg5594 — APPROVED** [review]

**Comment by lildude** [comment]
> I'm wondering if the “generic filetypes” feature added to Linguist in 2020 should accommodate “generic filenames” as well. @lildude, any thoughts?

I thought about this in https://github.com/github-linguist/linguist/pull/6364 and I'm still not sure it's worth the effort, but I'm open to being convinced with a compelling argument 😁.

> Any idea how we can get [our tests](https://github.com/github/linguist/actions/runs/4798736577/jobs/8537417513?pr=6391#step:5:68) to pass when both `hosts` _and_ `HOSTS` are expected to exist simultaneously under `samples/Hosts File/filenames/`…?

Fudge it 😁.

This simplest "solution" is skip the tests just for this language.

Are more convoluted "solution" is we adjust the various logic and use something like an underscore after these filenames, eg `hosts` and `HOSTS_`, to give these files unique names. 

Either way, I don't think we need to get too fancy right now as I don't expect this to be particularly common.

**Comment by Alhadis** [comment]
> Either way, I don't think we need to get too fancy right now as I don't expect this to be particularly common.

Yeah, me neither. 😀 I've added a comment with a permalink to your response (since this is the sort of last-minute hack that's all but indecipherable to my future self, give or take 2-3 years…)

**Comment by Alhadis** [comment]
@lildude I require your Rubyist expertise. Why has RubyGems suddenly carked it?

<details><summary>Output of <code>$ bundle exec rake samples</code></summary>

~~~markdown
--- ERROR REPORT TEMPLATE -------------------------------------------------------

```
Gem::GemNotFoundException: can't find gem bundler (= 2.4.5) with executable bundle
  /usr/local/lib/ruby/site_ruby/3.2.0/rubygems.rb:261:in `find_spec_for_exe'
  /usr/local/lib/ruby/site_ruby/3.2.0/rubygems.rb:241:in `bin_path'
  /usr/local/lib/ruby/site_ruby/3.2.0/bundler/rubygems_integration.rb:178:in `bin_path'
  /usr/local/lib/ruby/site_ruby/3.2.0/bundler/shared_helpers.rb:282:in `set_bundle_variables'
  /usr/local/lib/ruby/site_ruby/3.2.0/bundler/shared_helpers.rb:76:in `set_bundle_environment'
  /usr/local/lib/ruby/site_ruby/3.2.0/bundler/cli/exec.rb:20:in `run'
  /usr/local/lib/ruby/site_ruby/3.2.0/bundler/cli.rb:491:in `exec'
  /usr/local/lib/ruby/site_ruby/3.2.0/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /usr/local/lib/ruby/site_ruby/3.2.0/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
  /usr/local/lib/ruby/site_ruby/3.2.0/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
  /usr/local/lib/ruby/site_ruby/3.2.0/bundler/cli.rb:34:in `dispatch'
  /usr/local/lib/ruby/site_ruby/3.2.0/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
  /usr/local/lib/ruby/site_ruby/3.2.0/bundler/cli.rb:28:in `start'
  /usr/local/Cellar/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.10/exe/bundle:45:in `block in <top (required)>'
  /usr/local/lib/ruby/site_ruby/3.2.0/bundler/friendly_errors.rb:117:in `with_friendly_errors'
  /usr/local/Cellar/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.10/exe/bundle:33:in `<top (required)>'
  /usr/local/opt/ruby/bin/bundle:25:in `load'
  /usr/local/opt/ruby/bin/bundle:25:in `<main>'

```

## Environment

```
Bundler       2.4.5
  Platforms   ruby, x86_64-darwin-21
Ruby          3.2.2p53 (2023-03-30 revision e51014f9c05aa65cbf203442d37fef7c12390015) [x86_64-darwin-21]
  Full Path   /usr/local/opt/ruby/bin/ruby
  Config Dir  /usr/local/Cellar/ruby/3.2.2/etc
RubyGems      3.4.5
  Gem Home    /usr/local/lib/ruby/gems/3.2.0
  Gem Path    /Users/Alhadis/.gem/ruby/3.2.0:/usr/local/lib/ruby/gems/3.2.0:/usr/local/Cellar/ruby/3.2.2/lib/ruby/gems/3.2.0
  User Home   /Users/Alhadis
  User Path   /Users/Alhadis/.gem/ruby/3.2.0
  Bin Dir     /usr/local/lib/ruby/gems/3.2.0/bin
Tools         
  Git         2.40.1
  RVM         not installed
  rbenv       not installed
  chruby      not installed
Gem.ruby      /usr/local/opt/ruby/bin/ruby
bundle #!     /usr/local/Cellar/ruby/3.2.2/bin/ruby

```

## Bundler Build Metadata

```
Built At          2023-01-21
Git SHA           d8ff3b6e4a
Released Version  true
```

## Bundler settings

```
build.charlock_holmes
  Set for the current user (/Users/Alhadis/.bundle/config): "--with-icu-dir=/usr/local/opt/icu4c"
path
  Set for your local app (/Users/Alhadis/Forks/GitHub-Linguist/.bundle/config): "vendor/gems"
```

## Gemfile

### Gemfile

```ruby
source 'https://rubygems.org'
gemspec :name => "github-linguist"

group :debug do
  gem 'byebug' if RUBY_VERSION >= '2.2'
end
```

### Gemfile.lock

```
PATH
  remote: .
  specs:
    github-linguist (7.25.0)
      cgi
      charlock_holmes (~> 0.7.7)
      mini_mime (~> 1.0)
      rugged (~> 1.0)

GEM
  remote: https://rubygems.org/
  specs:
    addressable (2.8.4)
      public_suffix (>= 2.0.2, < 6.0)
    byebug (11.1.3)
    cgi (0.3.6)
    charlock_holmes (0.7.7)
    coderay (1.1.3)
    dotenv (2.8.1)
    faraday (2.7.4)
      faraday-net_http (>= 2.0, < 3.1)
      ruby2_keywords (>= 0.0.4)
    faraday-net_http (3.0.2)
    json (2.6.3)
    licensed (4.3.1)
      json (~> 2.6)
      licensee (~> 9.16)
      parallel (~> 1.22)
      pathname-common_prefix (~> 0.0.1)
      reverse_markdown (~> 2.1)
      ruby-xxHash (~> 0.4.0)
      thor (~> 1.2)
      tomlrb (~> 2.0)
    licensee (9.16.0)
      dotenv (~> 2.0)
      octokit (>= 4.20, < 7.0)
      reverse_markdown (>= 1, < 3)
      rugged (>= 0.24, < 2.0)
      thor (>= 0.19, < 2.0)
    method_source (1.0.0)
    mini_mime (1.1.2)
    minitest (5.18.0)
    mocha (1.16.1)
    nokogiri (1.15.0-x86_64-darwin)
      racc (~> 1.4)
    octokit (6.1.1)
      faraday (>= 1, < 3)
      sawyer (~> 0.9)
    parallel (1.23.0)
    pathname-common_prefix (0.0.1)
    plist (3.7.0)
    pry (0.14.2)
      coderay (~> 1.1)
      method_source (~> 1.0)
    public_suffix (5.0.1)
    racc (1.6.2)
    rake (13.0.6)
    rake-compiler (0.9.9)
      rake
    reverse_markdown (2.1.1)
      nokogiri
    ruby-xxHash (0.4.0.2)
    ruby2_keywords (0.0.5)
    rugged (1.6.3)
    sawyer (0.9.2)
      addressable (>= 2.3.5)
      faraday (>= 0.17.3, < 3)
    thor (1.2.2)
    tomlrb (2.0.3)
    yajl-ruby (1.4.3)

PLATFORMS
  x86_64-darwin-21

DEPENDENCIES
  bundler (~> 2.0)
  byebug
  github-linguist!
  licensed (~> 4.0)
  licensee (~> 9.15)
  minitest (~> 5.15)
  mocha (~> 1.3)
  plist (~> 3.1)
  pry (~> 0.14)
  rake (~> 13.0)
  rake-compiler (~> 0.9)
  yajl-ruby (~> 1.4)

BUNDLED WITH
   2.4.5
```

## Gemspecs

### github-linguist.gemspec

```ruby
require File.expand_path('../lib/linguist/version', __FILE__)

Gem::Specification.new do |s|
  s.name    = 'github-linguist'
  s.version = ENV['GEM_VERSION'] || Linguist::VERSION
  s.summary = "GitHub Language detection"
  s.description = 'We use this library at GitHub to detect blob languages, highlight code, ignore binary files, suppress generated files in diffs, and generate language breakdown graphs.'

  s.authors  = "GitHub"
  s.homepage = "https://github.com/github/linguist"
  s.license  = "MIT"
  s.metadata = {
    "github_repo" => "ssh://github.com/github/linguist"
  }

  s.files = Dir['{lib,ext}/**/*', 'grammars/*', 'LICENSE'] - Dir['lib/linguist/linguist.{so,bundle}']
  s.platform = Gem::Platform::RUBY
  s.executables = ['github-linguist', 'git-linguist']
  s.extensions = ['ext/linguist/extconf.rb']
  s.require_paths = ['lib', 'ext']

  s.add_dependency 'cgi',             '>= 0'
  s.add_dependency 'charlock_holmes', '~> 0.7.7'
  s.add_dependency 'mini_mime',       '~> 1.0'
  s.add_dependency 'rugged',          '~> 1.0'

  s.add_development_dependency 'minitest', '~> 5.15'
  s.add_development_dependency 'rake-compiler', '~> 0.9'
  s.add_development_dependency 'mocha', '~> 1.3'
  s.add_development_dependency 'plist', '~>3.1'
  s.add_development_dependency 'pry', '~> 0.14'
  s.add_development_dependency 'rake', '~> 13.0'
  s.add_development_dependency 'yajl-ruby', '~> 1.4'
  s.add_development_dependency 'licensed', '~> 4.0'
  s.add_development_dependency 'licensee', '~> 9.15'
  s.add_development_dependency 'bundler', '~> 2.0'
end
```

--- TEMPLATE END ----------------------------------------------------------------

Unfortunately, an unexpected error occurred, and Bundler cannot continue.

First, try this link to see if there are any existing issue reports for this error:
https://github.com/rubygems/rubygems/search?q=can%27t+find+gem+bundler+%28%3D+2.4.5%29+with+executable+bundle&type=Issues

If there aren't any reports for this error yet, please fill in the new issue form located at https://github.com/rubygems/rubygems/issues/new?labels=Bundler&template=bundler-related-issue.md, and copy and paste the report template above in there.
~~~

</details>

This is stopping me from running tests locally (hence the uncaught test failure above).

File: test/test_samples.rb [hunk 1]
```diff
@@ -90,6 +93,9 @@ def test_filename_listed
       end
 
       language.filenames.each do |filename|
+        # Kludge for an unusual edge-case; see https://bit.ly/41EyUkU
+        next if language.name == "Hosts File" && filename == "HOSTS"
```
Comment (lildude) [inline test/test_samples.rb:97]: You'll need to consider the "INI" entry too. How about:

```suggestion
        next if ["Host File", "INI"].include?(language.name) && filename == "HOSTS"
```

File: test/test_samples.rb [hunk 2]
```diff
@@ -46,6 +46,9 @@ def test_filename_listed
     Samples.each do |sample|
       if sample[:filename]
         listed_filenames = Language[sample[:language]].filenames
+        if sample[:language] == "Hosts File" # Kludge for https://bit.ly/41EyUkU
+          listed_filenames -= ["HOSTS"]
+        end
```
Comment (lildude) [inline test/test_samples.rb:51]: I think we need to consider the INI lanugage here too.

How about:

```suggestion
        listed_filenames -= ["HOSTS"] if ["Hosts File", "INI"].include?(sample[:language])
```

**Review by lildude — COMMENTED** [review]
See inline comments

**Comment by Alhadis** [comment]
> I therefore wonder if `^((\d{1,3}\.){3,}\d{1,3}|[0-9a-f:]{1,5}:([0-9a-f:]{1,34})?) ` would work better.

@DandelionSprout I've already started work on improving the accuracy of matching IP addresses in `HOSTS` files; it's currently staged locally. I hope to get it done before this PR's merged and/or @lildude cuts the next release.

**Review by lildude — APPROVED** [review]

**Comment by DandelionSprout** [comment]
So, now that this has been merged, which colours will end up being used in Hosts files and for what, and is there a timeline for when such colours would begin to appear in GitHub's web GUI?
